### PR TITLE
📲 feat: post endpoints

### LINKED ISSUE #70: [BE] post endpoints
# **Post Endpoints**
As a user, I want to make a post in the group I'm a member of to share my daily life with family and friends.

---

## **Acceptance Criteria**
POST /api/v1/groups/:id/posts/: Create a post
GET /api/v1/groups/:groupId/posts/:postId: Get a post's detail
PATCH /api/v1/groups/:id/posts/:postId: Update the post's detail
DELETE /api/v1/groups/:id: Delete a post

- [x] Define a post schema and endpoints in the openapi.yaml schema.
- [x] Define the route for posts and create the proper type aliases from the schema generation.
- [x] Define transaction interface and its implementation to work with the database.
- [x] Define service interface to properly handle core business logic together as well as delegate to transaction layer.
- [x] Define controller interface, properly type checking with the previously defined type aliases from schema. <br>

[Backend Guidelines](https://github.com/GenerateNU/dearly/blob/main/docs/BACKEND.md)
[Database Drizzle Schema](https://github.com/GenerateNU/dearly/blob/main/backend/src/entities/schema.ts)
[Database Supabase Visualization](https://supabase.com/dashboard/project/wtmsxzaucgdwttgwpmjy/database/schemas)

> [!NOTE]
> - only members should be able to make a post in the group
> - only user who creates the post can update or delete it

Things to think about (what kind of HTTP error would you throw in these cases?):
- What if a user is not in a group and try to make a post?
- What if someone tries to delete or update a post when they are not the users who post it?

---

## **Testing Plan**

- Brainstorm different scenarios and edge cases and what kind of HTTP status code would be returned in each one.
- Ensure that the controller returns the proper response codes depending on proper requests.
- Ensure that the API returns the exact schema as defined in the OpenAPI.yaml file, this includes error responses as well.
- Testing a full cycle would also be great as well.

---

## **Definition of Done**

1. All acceptance criteria are met.  
2. All tests pass locally and cover all edge cases.

### PR DESCRIPTION
**Pull request by in-mai-space**
# Description

Closes #70 
- Allow users to create, get, update and delete a post (with media field that are arrays of image or video url)
- Enforce permission on the endpoints:
  - only group members can create a post
  - only members of the group can get a post in the group
  - only the owner of a post can update and delete that post

Tests are added to check whether the endpoints:
- Return correct status code
- Return correct error message based on group membership / post ownership
- Reject payload with bad request values (enums, min & max limit)
- Reject bad UUID in route params

## Checklist:

- [x] I have performed a self-review of my code.
- [x] I have added necessary tests (unit, integration, etc.) to cover the new functionality or fix.
- [x] All new and existing tests pass locally.
- [x] I have updated the documentation to reflect changes (e.g., README, code comments).
- [x] I have removed any commented-out code, debug statements, or unnecessary console logs.

### CODE REVIEW

File: backend/src/tests/helpers/test-constants.ts [hunk 1]
```diff
@@ -1,7 +1,15 @@
...
 export const USER_ALICE_ID = "00000000-0000-0000-0000-000000000000";
-export const USER_BOB_ID = "00000000-0000-0000-0000-000000000001";
-export const USER_ANA_ID = "00000000-0000-0000-0000-000000000002";
+export const USER_BOB_ID = generateUUID();
+export const USER_ANA_ID = generateUUID();
+export const DEARLY_GROUP_ID = generateUUID();
+export const POST_ID = generateUUID();
+export const MEDIA_ONE_ID = generateUUID();
+export const MEDIA_TWO_ID = generateUUID();
+export const MEDIA_THREE_ID = generateUUID();
```
Comment (stoneliuCS) [inline backend/src/tests/helpers/test-constants.ts:12]: Unsure how uuids are being used exactly around here, but this seems to be a little bit of indeterminism here with generating new uuids and making them a global variable every test run. (This might not be a huge deal depending on how the tests are constructed, but generally try to stray away from any non deterministic behavior in testing unless they are property tests)

File: backend/src/tests/posts/delete.test.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+      },
+    ],
+  };
+
+  const ALICE_JWT = generateJWTFromID(USER_ALICE_ID);
+  const BOB_JWT = generateJWTFromID(USER_BOB_ID);
+  const ANA_JWT = generateJWTFromID(USER_ANA_ID);
+
+  beforeAll(async () => {
+    app = await startTestApp();
```
Comment (stoneliuCS) [inline backend/src/tests/posts/delete.test.ts:36]: A maybe useful (or maybe not) abstraction is to define setup and teardown functions that each test can take advantage of. For example, before each test we can run a post request so we have some initial data in our database to play with then remove any lingering data left in the test database. Although sharing one instance of the database sounds reasonable as well, this method can make each testing environment act independently from each other (Which may prevent weird behavior in the order of which tests are executed)

**Review by stoneliuCS — COMMENTED** [review]
Excellent work as always Mai!

**Review by stoneliuCS — APPROVED** [review]

File: backend/src/tests/posts/delete.test.ts (hunk 2 above)
Comment (in-mai-space) [reply backend/src/tests/posts/delete.test.ts:36]: The `startTestApp` will reset the database before each test suite, so there’s no conflict between tests regarding the data present. I initially tried using `beforeEach` to reset the database before every test case, but since there are a lot of tests, PostgreSQL eventually threw an error due to too many client connection attempts. Doing this will make each test connect to the database, even for tests like a bad request that will never reach the transaction layer, which can also slow down each test run since each test case needs to connect to PostgreSQL.

Yeh agree we need some initial data for testing. Currently, I insert mock data into the database with a specific UUID to test the post endpoint independently without needing to create the associated user, group, and member data.
